falcon/asgi/ws.py

```diff
@@ -0,0 +1,562 @@
+            #   the "websockets" library that allows exceptions to bubble up,
+            #   we will have an error raised on client disconnect.
+            #
+            #   Daphne, on the other hand, does not raise an error but just
+            #   eats the message. This approach is actually more in keeping
+            #   with the ASGI spec, but poses its own challenges.
+
+            translated_ex = self._translate_webserver_error(ex)
+            if translated_ex:
+                raise translated_ex
+
+            # NOTE(kgriffs): Re-raise other errors directly so that we don't
+            #   obscure the traceback.
+            raise
+
+    async def _receive(self) -> dict:
+        event = await self._asgi_receive()
+
+        event_type = event['type']
+
+        if event_type != EventType.WS_RECEIVE:
+            # NOTE(kgriffs): Based on the ASGI spec, there are no other
+            #   event types that should be emitted by the protocol server,
+            #   but we sanity-check it here just in case.
+            assert event_type == EventType.WS_DISCONNECT
+
+            self._state = _WebSocketState.CLOSED
+            self._close_code = event.get('code', WSCloseCode.NORMAL)
+            raise errors.WebSocketDisconnected(self._close_code)
+
+        return event
+
+    def _require_accepted(self):
+        if self._state == _WebSocketState.HANDSHAKE:
+            raise errors.OperationNotAllowed('WebSocket connection has not yet been accepted')
+        elif self._state == _WebSocketState.CLOSED:
+            raise errors.WebSocketDisconnected(self._close_code)
+
+    def _translate_webserver_error(self, ex):
+        s = str(ex)
+
+        # NOTE(kgriffs): uvicorn or any other server using the "websockets"
+        #   package that allows exceptions to bubble up
+        if 'code = 1000 (OK)' in s:
+            return errors.WebSocketDisconnected(WSCloseCode.NORMAL)
+
+        # NOTE(kgriffs): Autobahn (used by Daphne) raises a generic exception
+        #   with this message
+        if 'protocol accepted must be from the list' in s:
+            return ValueError('WebSocket subprotocol must be from the list sent by the client')
+
+        return None
+
+
+class WebSocketOptions:
+    __slots__ = ['error_close_code', 'max_receive_queue', 'media_handlers']
+
+    def __init__(self):
+        self.media_handlers = {
+            WebSocketPayloadType.TEXT: media.JSONHandlerWS(),
+            WebSocketPayloadType.BINARY: media.MessagePackHandlerWS(),
+        }
+
+        # Internal Error
+        #
+        #   See also: https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent
+        #
+        self.error_close_code = WSCloseCode.SERVER_ERROR
+
+        # NOTE(kgriffs): The websockets library itself will buffer, so we keep
+        #   this value fairly small to reduce memory usage by default. On the
+        #   other hand, AFAICT, autobahn-based servers such as Daphne do not
```

Comment:
It might be just me, but I'm not sure I'm following the reasoning here for the Daphne part.
Why does it help to increase the value on the Falcon side? Wouldn't it add even more buffer bloat to the system?
Or does it make sense by accepting Daphne's reality in order to avoid queue full errors when Daphne is trying to spill a huge buffer onto our app? (Although it seems we are instead deferring work to a future?)

Reply:
Actually, you are right. I'm not sure what I was thinking when I wrote this. Even if we block on our side, the WebSocket library will continue to build up its own buffer, so back pressure is not really something we can affect. Supposing the app sporadically slows down its consumption and the Daphne buffer builds up, when the app speeds up, having a larger Falcon buffer could potentially reduce the number of pauses in the future running the pump as it absorbs the backlog. But its hard to guess if such pauses would materially impact performance, and even so it would require comprehensive workload/app-specific benchmarking.

Reply:
Perhaps I should document this after all, and mention that setting it to zero will disable buffering?

Reply:
+1 for making buffering configurable, and allowing to disable it altogether.

Not sure what the default value should be... Gut feeling says, given that most of ASGI stuff seems to err on the side of excessive buffer bloat, we might want to keep this disabled by default? I would need to test how this feels in practical use tho.

Reply:
Having it enabled leads to a better developer experience IMO since it abstracts away subtleties re detecting when the client disconnects.

Reply:
I suppose we could disable it if we detect uvicorn but the danger is that they change their behavior at some point to more strictly observe the ASGI spec.

Reply:
Ah, yeah, I completely forgot about that aspect of the spec 🤔
Agreed that we must handle it on our side because of that.